### PR TITLE
IRI regex (RFC3987)

### PR DESCRIPTION
**Pull request by iTakeshi**
The first step to solve https://github.com/endaaman/tym/issues/30.
The implementation follows the Backus-Naur definition in [RFC3987](http://www.faqs.org/rfcs/rfc3987.html).
Basic functionalities are already working but there remain some critical bugs.
I'm still working on them, but for the moment I submit this PR in order to avoid duplicate works.

### CODE REVIEW

**Comment by endaaman** [comment]
Great work! We need this. Please keep it up.

**Comment by iTakeshi** [comment]
Note: we should be aware of the [new URL standard](https://url.spec.whatwg.org) by WHATWG, which aims to deprecate RFC3987. However it does not contain comprehensive grammer (its parser is defined as a set of procedures), so using RFC3987 would be OK for now.

File: src/app.c [hunk 1]
```diff
@@ -173,6 +179,17 @@ void on_activate(GApplication* app, void* user_data)
   VteTerminal* vte = context_get_vte(context);
   GtkWindow* window = context_get_window(context);
 
+  VteRegex* regex = vte_regex_new_for_match(IRI, strlen(IRI), PCRE2_UTF, &error);
```
[the remark this answers is not in the document]
Comment (iTakeshi) [reply src/app.c]: This warning comes from https://github.com/GNOME/vte/blob/master/src/vtegtk.cc#L1964 .
I'll add the flag.

File: src/app.c [hunk 2]
```diff
@@ -7,7 +7,11 @@
  * of the MIT license. See the LICENSE file for details.
  */
 
+#define PCRE2_CODE_UNIT_WIDTH 8
+#include <pcre2.h>
```
[the remark this answers is not in the document]
Comment (iTakeshi) [reply src/app.c]: Moved in the upcoming commit.

**Comment by iTakeshi** [comment]
Updated with some bug fixes and improvements.
I believe this works for a majority of cases, but there remains some bugs and limitations:
1. query strings `?foo=bar&baz=qux` sometimes aren't recognized, especially when there are multiple fields,
2. unwanted close parenthesis are matched when URLs are in parenthesis/braces/etc. i.e. if we have a markdown link `[link](https://example.com)`, the matched string will be `https://example.com)`.
For (1), it must be a bug in the IRI regex, but I can't see where it is.
For (2), this isn't a bug. If we really want to deal with it, we have to modify IRI definition.

Aren't PRs the target of CI? I checked updated Dockerfile on my PC, but you may want CI to cover PRs.

**Comment by iTakeshi** [comment]
Also, I'm currently working on the cursor. After that commit this PR would be ready.
I'll continue to work on click actions, but it will be another PR.

**Comment by endaaman** [comment]
> you may want CI to cover PRs.

I turned on `Build forked pull requests` in CI settings page. Next commit should be tested by CI.

> but there remains some bugs and limitations:

You mean theses problems are not fixed in this PR (and new issues about them should be opened) ?

> I'll continue to work on click actions, but it will be another PR.

OK.

**Comment by iTakeshi** [comment]
I'm now working on query string bug, but parenthesis handling goes to another issue.
Before this PR become ready:
- [ ] fix query string detection
- [ ] change cursor shape on hover

**Comment by iTakeshi** [comment]
Implemented cursor change, and now this PR is ready for your review.

**Comment by endaaman** [comment]
Seems good. Thank you very much!

**Pull request merged**